### PR TITLE
Added GreaterThan version update check to catalog api

### PR DESCRIPTION
**Pull request by Skarlso**
### Description

Related PCTL PR https://github.com/weaveworks/pctl/pull/84/files
Partially deals with https://github.com/weaveworks/profiles/issues/101

### Checklist
- [X] Added tests that cover your change
- [X] Added/modified documentation as required (such as the `README.md` (diagrams, usage, roadmap etc), and anything in `examples/`)
- [X] Made sure the title of the PR is a good description that can go into the release notes
- [X] Added a `kind` label to the PR (e.g. `kind/feature`)

### CODE REVIEW

File: pkg/catalog/catalog.go [hunk 1]
```diff
@@ -71,7 +73,11 @@ func (c *Catalog) GetWithVersion(sourceName, profileName, profileVersion string)
 	}
 
 	if profileVersion == "latest" {
-		return getLatestVersion(profiles.([]profilesv1.ProfileDescription), profileName)
+		versions := c.GetGreaterThan(sourceName, profileName, profileVersion)
+		if len(versions) == 0 {
+			return nil
+		}
+		return versions[0].DeepCopy()
```
Comment (aclevername) [inline pkg/catalog/catalog.go:80]: nice re-use :ok_hand:

File: pkg/catalog/catalog.go [hunk 2]
```diff
@@ -5,7 +5,9 @@ import (
 	"strings"
 	"sync"
 
-	"github.com/hashicorp/go-version"
+	"github.com/Masterminds/semver/v3"
+	"github.com/fluxcd/pkg/version"
+
```
Comment (aclevername) [inline pkg/catalog/catalog.go:10]: I'm OOL, was there actually anything wrong with the previous version parser? We use it in eksctl and it works well :shrug:

Comment (Skarlso) [reply pkg/catalog/catalog.go:10]: This is an attempt at unifying the work we do as a company. /cc @bigkevmcd 
We can skip it if you think that it should not be included ATM.

File: pkg/catalog/catalog.go [hunk 3]
```diff
@@ -84,20 +90,29 @@ func (c *Catalog) GetWithVersion(sourceName, profileName, profileVersion string)
...
+	}
+	for _, p := range profiles.([]profilesv1.ProfileDescription) {
+		v, err := version.ParseVersion(p.Version)
 		if err != nil {
 			continue
 		}
-
 		if p.Name == profileName {
-			profilesWithValidVersion = append(profilesWithValidVersion, profileDescriptionWithVersion{profileDescription: p, semverVersion: v})
+			if profileVersion == "latest" || (cv != nil && v.GreaterThan(cv)) {
```
Comment (aclevername) [inline pkg/catalog/catalog.go:113]: do we need the nil check? looks smelly :laughing:

File: pkg/catalog/catalog_test.go [hunk 4]
```diff
@@ -83,4 +82,19 @@ var _ = Describe("Catalog", func() {
 			})
 		})
 	})
+
+	Describe("GetGreaterThan", func() {
+		It("lists all available versions which are greater than the current version", func() {
```
Comment (aclevername) [inline pkg/catalog/catalog_test.go:87]: nit: lets call out in the code & test its an ordered list, to make sure folks know :smile:

File: pkg/catalog/catalog.go (hunk 3 above)
Comment (Skarlso) [reply pkg/catalog/catalog.go:113]: We don't necessarily need it, but the static check is happier this way. :)

That's because the latest check above might turn the cv into a nil pointer. But that wouldn't be used anyways, since profileVersion == "latest"... But the static checker is still complaining about it.

Comment (Skarlso) [reply pkg/catalog/catalog.go:113]: Oh no, actually I changed this recently and didn't update. :D :D

File: pkg/api/api.go [hunk 5]
```diff
@@ -38,6 +40,7 @@ func New(profileCatalog Catalog) API {
 	r.HandleFunc("/profiles", a.ProfilesHandler)
 	r.HandleFunc("/profiles/{catalog}/{profile}", a.ProfileHandler)
 	r.HandleFunc("/profiles/{catalog}/{profile}/{version}", a.ProfileWithVersionHandler)
+	r.HandleFunc("/profiles/{catalog}/{profile}/{version}/update", a.ProfileGreaterThanVersionHandler)
```
Comment (aclevername) [inline pkg/api/api.go:43]: Sorry in advance D: 

I know we are picking one and running, but `update` just seems confusing, mainly from a CRUD point of view too. Wb `newer` or `available_updates` or something that doesn't make me think its updating the profile :joy: . It reads slightly cleaner:
```
/profiles/nginx-catalog/weaveworks-nginx/v0.1.0/update
vs
/profiles/nginx-catalog/weaveworks-nginx/v0.1.0/newer
/profiles/nginx-catalog/weaveworks-nginx/v0.1.0/newer_versions
/profiles/nginx-catalog/weaveworks-nginx/v0.1.0/available_updates

```

File: pkg/catalog/catalog.go (hunk 2 above)
Comment (aclevername) [reply pkg/catalog/catalog.go:10]: nah thats fine

File: pkg/api/api.go [hunk 6]
```diff
@@ -70,6 +73,17 @@ func (a *API) ProfileWithVersionHandler(w http.ResponseWriter, r *http.Request)
...
+func (a *API) ProfileGreaterThanVersionHandler(w http.ResponseWriter, r *http.Request) {
+	sourceName, profileName, catalogVersion := mux.Vars(r)["catalog"], mux.Vars(r)["profile"], mux.Vars(r)["version"]
+	result := a.profileCatalog.GetGreaterThan(sourceName, profileName, catalogVersion)
+	if result == nil {
+		w.WriteHeader(404)
+		return
+	}
+	marshalResponse(w, result)
+}
+
```
Comment (aclevername) [inline pkg/api/api.go:86]: do we need to check for empty strings? This might apply to other funcs in here too

File: pkg/catalog/catalog.go [hunk 7]
```diff
@@ -71,7 +73,11 @@ func (c *Catalog) GetWithVersion(sourceName, profileName, profileVersion string)
 	}
 
 	if profileVersion == "latest" {
-		return getLatestVersion(profiles.([]profilesv1.ProfileDescription), profileName)
+		versions := c.GetGreaterThan(sourceName, profileName, profileVersion)
```
Comment (aclevername) [inline pkg/catalog/catalog.go:76]:
```suggestion
		versions := c.ProfileGreaterThanVersion(sourceName, profileName, profileVersion)
```

for consistency with the handler? its also more descriptive :smile:

Comment (Skarlso) [reply pkg/catalog/catalog.go:76]: I'm not sure I follow this one? This is inside catalog, has nothing to do with the handler. We are consistent with the names in this file and not another. :D

File: pkg/catalog/catalog.go [hunk 8]
```diff
@@ -84,20 +90,29 @@ func (c *Catalog) GetWithVersion(sourceName, profileName, profileVersion string)
 
 type profileDescriptionWithVersion struct {
 	profileDescription profilesv1.ProfileDescription
-	semverVersion      *version.Version
+	semverVersion      *semver.Version
 }
 
-func getLatestVersion(profiles []profilesv1.ProfileDescription, profileName string) *profilesv1.ProfileDescription {
+// GetGreaterThan returns all profiles which are of a greater version for a given profile with a version.
```
Comment (aclevername) [inline pkg/catalog/catalog.go:96]:
```suggestion
// GetGreaterThan returns all profiles which are of a greater version for a given profile with a version.
// If set to "latest" all versions are returned. Versions are ordered in descending order
```

File: pkg/catalog/catalog_test.go [hunk 9]
```diff
@@ -58,16 +58,15 @@ var _ = Describe("Catalog", func() {
 
 		When("version is set to latest", func() {
 			It("returns the latest version", func() {
-				profiles := []profilesv1.ProfileDescription{{Name: "foo", Version: "v0.1.0"}, {Name: "foo", Version: "0.2.0"}, {Name: "bar", Version: "0.3.0"}, {Name: "foo"}}
+				profiles := []profilesv1.ProfileDescription{{Name: "foo", Version: "v0.1.0"}, {Name: "foo", Version: "v0.2.0"}, {Name: "bar", Version: "v0.3.0"}, {Name: "foo"}}
```
Comment (aclevername) [inline pkg/catalog/catalog_test.go:61]: I think we purposely had a mix of `v` and without, just for sanity checking :smile:

File: pkg/api/api.go (hunk 6 above)
Comment (Skarlso) [reply pkg/api/api.go:86]: Sure, why not. :)

File: pkg/catalog/catalog_test.go (hunk 9 above)
Comment (Skarlso) [reply pkg/catalog/catalog_test.go:61]: ah, I see. :D Okay :D

File: pkg/catalog/catalog_test.go [hunk 10]
```diff
@@ -83,4 +82,19 @@ var _ = Describe("Catalog", func() {
...
+			c.Update(catName, profiles...)
+
+			Expect(c.GetGreaterThan(catName, "foo", "v0.1.0")).To(Equal(
+				[]profilesv1.ProfileDescription{
+					{Name: "foo", Version: "v0.3.0", CatalogSource: catName},
+					{Name: "foo", Version: "v0.2.0", CatalogSource: catName},
+				},
+			))
+		})
+	})
```
Comment (aclevername) [inline pkg/catalog/catalog_test.go:99]: I feel burned by my failed attempt at this in the past. shall we expand the test to have:
a version "older" than `v0.1.0` (and see that not returned) and a version newer than `v0.1.0` belongs to a different profile? Just to cover all the edge cases. This way I can't break it in the future :smile:

Comment (Skarlso) [reply pkg/catalog/catalog_test.go:99]: Hahahaha, yeah sure. :)

File: pkg/catalog/catalog.go (hunk 7 above)
Comment (aclevername) [reply pkg/catalog/catalog.go:76]: I meant more that `GetGreaterThan` inside `catalog` package is sort of vague.. greater than what? `ProfilesGreaterThanVersion` is much cleaner on what it returns

Comment (Skarlso) [reply pkg/catalog/catalog.go:76]: Oh, ooooh I see. I left out version there. :D Why did I do that? :D

File: pkg/api/api.go (hunk 6 above)
Comment (Skarlso) [reply pkg/api/api.go:86]: Let's skip that because that means changes in PCTL to properly show a 400. Because 404 is not very valid in case parameters are incorrect.
Created https://github.com/weaveworks/profiles/issues/140 to track

Comment (aclevername) [reply pkg/api/api.go]: :+1:

File: pkg/catalog/catalog_test.go [hunk 11]
```diff
@@ -83,4 +82,26 @@ var _ = Describe("Catalog", func() {
...
+				{Name: "foo", Version: "v0.3.0"}, {Name: "foo"},
+				{Name: "foo2", Version: "v0.3.0"}, {Name: "foo"},
+				{Name: "foo2", Version: "v0.3.1"}, {Name: "foo"},
+			}
+			c.Update(catName, profiles...)
+
+			Expect(c.ProfilesGreaterThanVersion(catName, "foo", "v0.1.0")).To(Equal(
+				[]profilesv1.ProfileDescription{
+					{Name: "foo", Version: "v0.3.0", CatalogSource: catName},
+					{Name: "foo", Version: "v0.2.0", CatalogSource: catName},
```
Comment (aclevername) [inline pkg/catalog/catalog_test.go:102]: nice

**Review by aclevername — APPROVED** [review]
LGTM :smile:
